numba/typedobjectutils.py

```diff
@@ -39,18 +40,34 @@ def _sentry_safe_cast(fromty, toty):
     """Check and raise TypingError if *fromty* cannot be safely cast to *toty*
     """
     tyctxt = cpu_target.typing_context
+    fromty, toty = map(types.unliteral, (fromty, toty))
     by = tyctxt.can_convert(fromty, toty)
+
+    def warn():
+        m = 'unsafe cast from {} to {}. Precision maybe lost.'
+        warnings.warn(m.format(fromty, toty),
+                      category=NumbaTypeSafetyWarning)
+
+    isint = lambda x: isinstance(x, types.Integer)
+    isflt = lambda x: isinstance(x, types.Float)
+    # Only check against numeric types.
     if by is None or by > Conversion.safe:
-        if isinstance(fromty, types.Integer) and isinstance(toty, types.Integer):
+        if isint(fromty) and isint(toty):
             # Accept if both types are ints
-            return
-        if isinstance(fromty, types.Integer) and isinstance(toty, types.Float):
+            warn()
+        elif isint(fromty) and isflt(toty):
             # Accept if ints to floats
-            return
-        if isinstance(fromty, types.Float) and isinstance(toty, types.Float):
+            warn()
+        elif isflt(fromty) and isflt(toty):
             # Accept if floats to floats
-            return
-        raise TypingError('cannot safely cast {} to {}'.format(fromty, toty))
+            warn()
+        elif not isinstance(toty, types.Number):
+            # Non-numbers
+            warn()
```

Comment:
I think `(complex64, complex128)` casting should work ok but warn and not be a hard error?

Reply:
good catch. fixe in https://github.com/numba/numba/pull/4330/commits/91aef6bfff8d03d0d22d65745e7732a15611d056